include/pybind11/pybind11.h

```diff
@@ -69,6 +69,13 @@ inline bool apply_exception_translators(std::forward_list<ExceptionTranslator> &
     return false;
 }
 
+inline bool is_function_record_capsule(const capsule &cap) {
+    if (cap.name() == nullptr) {
+        return false;
+    }
+    return (std::strcmp(cap.name(), function_record::function_capsule_name) == 0);
```

Comment:
The primary thing we are worried about isn't speed, it's ABI safety.

The question is what function_record's should pybind11 try to manipulate?

Options are:
1. function_records only from the same .so (pointer comparison using local_internals)
2. function_records only from the same PYBIND11_INTERNALS_ID (pointer comparison using internals)
3. all function_records (std::strcmp using manual string)

See my response to the above discussion for what "manipulation" pybind11 does (but note that the exact type of manipulation probably doesn't matter that much for ABI; any use is UB and a bad idea).

Reply:
I would argue that 1 is safer than 2, and 2 is safer than 3.

Reply:
Thanks a lot for the explanation! I'll need a little bit of time to look more carefully. But one quick high-level question: do we have timings, even just crude, for what the performance hit is for a "typical situation in the wild" (vs. micro-benchmark only)?

A super crude idea (not sure it's feasible & meaningful, but just to start somewhere): run certain existing unit tests always returning `false` here.

Reply:
The performance benefit of functional.h's optimization will be extremely dependent on the setup. It's really just the cost of the type_converters for each input argument and return value. For something like std::vector, it will be very high, for something like integers it's basically zero.

The bigger question is to what degree do applications rely on this optimization across pybind11 modules ...

Our unit tests only try this optimization for functions that take integers so there is practically performance difference (well, 170 nanoseconds).

Reply:
Thanks, that's very clear!

Hm ... I'd think people generally expect no significant performance regressions, that would rule out 1.

2. The internals version trick has known limitations and pitfalls (potential UB) that are generally accepted.

3. means modules are not ABI compatible, presumably not expected to cooperate, presumably have very little expectations re performance, apart maybe "it used to be faster" for a handful of users (who may be blissfully unaware of the functional.h mechanics).

I'm making this up, I don't think this can be decided in a pure scientific way.

I'd settle on 2.

Reply:
2 is now implemented.